### PR TITLE
Adds More checks on node request/usage

### PR DESCRIPTION
**Pull request by Kuranes**
Adds More checks on node request/usage
Adds varyings check between shader stage
Adds Comments
Adds new Kind of variables: Outputs (glFragColor, glPointsize, glPosition)

### CODE REVIEW

File: sources/osgShader/Compiler.js [hunk 1]
```diff
@@ -183,11 +186,43 @@ define( [
...
+            var cacheID = n.getID();
+            this._activeNodeList[ cacheID ] = n;
+            return n;
+        },
+        // during compilation we pop
+        // all node we do encounter
+        // so that we can warn about
+        // "leftover" once compilation
+        // is finished
+        tapNodeAsCompiled: function ( n ) {
```
Comment (cedricpinson) [inline sources/osgShader/Compiler.js]: is it to say markNodeAsUsed/Compiled ?

File: sources/osgShader/Compiler.js [hunk 2]
```diff
@@ -853,6 +890,9 @@ define( [
                 }
             }
             functor.call( functor, node );
+
+            // keep trace we visited.
+            this.tapNodeAsCompiled( node );
```
Comment (cedricpinson) [inline sources/osgShader/Compiler.js]: ok maybe markNodeAsVisited ?

File: tests/osgShader/Compiler.js [hunk 3]
```diff
@@ -78,7 +78,7 @@ define( [
                     value.apply( instance );
                     var t = instance.getType();
 
-                    if ( instance._name === 'Variable' ) {
+                    if ( instance._name === 'Variable' || instance._unique ) {
```
Comment (cedricpinson) [inline tests/osgShader/Compiler.js]: we should use accessors getName and isUnique

File: sources/osgShader/Compiler.js (hunk 1 above)
Comment (Kuranes) [inline sources/osgShader/Compiler.js]: yes, but as we may encounter multiple times the node (shadergraph is visited for extension, defines, globalfunc, ) I thought "mark" was a bit "definitive" whereas it a Node can be tapped several times

File: sources/osgShader/Compiler.js (hunk 2 above)
Comment (Kuranes) [inline sources/osgShader/Compiler.js]: markNodeAsVisitedAtLeastOnce ?

File: tests/osgShader/Compiler.js (hunk 3 above)
Comment (Kuranes) [inline tests/osgShader/Compiler.js]: k.

File: sources/osgShader/Compiler.js [hunk 4]
```diff
@@ -183,11 +186,45 @@ define( [
...
+        // "leftover" once compilation
+        // is finished
+        // Note: same node may be marked multiple time
+        // do not use it as a "once and for all mark thing"
+        markNodeAsVisited: function ( n ) {
+            var cacheID = n.getID();
+            if ( this._activeNodeList[ cacheID ] === n ) {
+                this._compiledNodeList[ cacheID ] = n;
+            } else {
+                Notify.error( 'Node not requested by using Compiler getNode and/or not registered in nodeFactory' );
```
Comment (cedricpinson) [inline sources/osgShader/Compiler.js]: maybe it's not an error but a warning ?

File: sources/osgShader/Compiler.js [hunk 5]
```diff
@@ -1239,30 +1290,55 @@ define( [
...
+            // Process defines, add precision, resolve include pragma
             var shader = this._shaderProcessor.processShader( shaderStr, defines, extensions, type );
 
+            // Check
+            var compiledNodes = Object.keys( this._compiledNodeList );
+            var activeNodes = Object.keys( this._activeNodeList );
+            activeNodes.filter( function ( i ) {
+                var found = compiledNodes.indexOf( i ) !== -1;
+                if ( !found ) {
+                    Notify.log( 'Nodes Not Used: ' + i + ' ' + this._activeNodeList[ i ].getName() );
```
Comment (cedricpinson) [inline sources/osgShader/Compiler.js]: log or warning ?

**Comment by cedricpinson** [comment]
That will help for sure, warning / error / log, could you check what's more appropriate ?

File: sources/osgShader/Compiler.js (hunk 4 above)
Comment (Kuranes) [inline sources/osgShader/Compiler.js]: true, could be a warning.

File: sources/osgShader/Compiler.js (hunk 5 above)
Comment (Kuranes) [inline sources/osgShader/Compiler.js]: warning indeed.

**Comment by Kuranes** [comment]
Agree, will make a check

**Comment by Kuranes** [comment]
reworked log to warning + better warning message.
Maybe We should use develblock so that those only appear/compute in debug build ?

File: sources/osgShader/node/data.js [hunk 6]
```diff
@@ -126,12 +130,19 @@ define( [
...
+    // Derive from that for new outputs
+    // gl_FragDepth, etc.
+    var Output = function ( type, wholeName ) {
+        Variable.call( this, type, wholeName );
     };
-    glFragColor.prototype = MACROUTILS.objectInherit( Variable.prototype, {
 
+    Output.prototype = MACROUTILS.objectInherit( Variable.prototype, {
+        _unique: true,
+        getUnique: function () {
```
Comment (cedricpinson) [inline sources/osgShader/node/data.js]: isUnique

**Pull request merged**